File: 0739-daily-temperatures/0739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:

        res = [0] * len(temperatures)
        stack = []

        for i in range(len(temperatures)):
            c_temp = temperatures[i]
            while stack and stack[-1][0] < c_temp:
                temp, idx = stack.pop()
                res[idx] = i - idx
            
            stack.append([c_temp, i])
        
        return res

        

File: 0739-daily-temperatures/0739-daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        stack<pair<int, int>> stk;
        int n = temperatures.size();
        vector<int> t = temperatures;

        vector<int> result(n);

        for (int i = 0; i < n; i++) {
            while (!stk.empty() && stk.top().second < t[i]) {
                int prevDay = stk.top().first;
                stk.pop();
                result[prevDay] = i - prevDay;
            }
            stk.push({i, t[i]});
        }
        return result;
    }
};